trin-core/src/portalnet/overlay.rs

```diff
@@ -253,35 +253,54 @@ where
             ));
         }
 
-        // TODO: Verify overlay content data with an Oracle
-
-        // Temporarily store content key/value pairs to propagate here
-        let mut content_keys_values: Vec<(TContentKey, ByteList)> = Vec::new();
-
-        // Try to store the content into the database and propagate gossip the content
         for (content_key, content_value) in content_keys.into_iter().zip(content_values.to_vec()) {
             match TContentKey::try_from(content_key) {
                 Ok(key) => {
-                    // Store accepted content in DB
-                    self.store_overlay_content(&key, content_value.clone());
-                    content_keys_values.push((key, content_value))
+                    // Spawn a task that...
+                    // - Validates accepted content (this step requires a dedicated task since it
+                    // might require non-blocking requests to this/other overlay networks)
+                    // - Checks if validated content should be stored, and stores it if true
+                    // - Propagate all validated content
+                    let validator = Arc::clone(&self.validator);
+                    let storage = Arc::clone(&self.storage);
+                    let kbuckets = Arc::clone(&self.kbuckets);
+                    let request_tx = self.request_tx.clone();
+
+                    tokio::spawn(async move {
+                        // Validated received content
+                        validator
+                            .validate_content(&key, &content_value.to_vec())
+                            .await
+                            // Skip storing & propagating content if it's not valid
+                            .expect("Unable to validate received content: {err:?}");
+
+                        // Check if data should be stored, and store if true.
+                        // Ignore error since all validated content is propagated.
+                        let _ = storage
+                            .write()
+                            .store_if_should(&key, &content_value.to_vec());
+
+                        // Propagate all validated content, whether or not it was stored.
+                        Self::propagate_gossip((key, content_value), kbuckets, request_tx);
```

Comment:
I definitely agree one offer message with many key/value > many messages with one key/value. 

It seems we have 3 goals here, and one of them will need to sacrificed to achieve the other two.
- Only `OFFER` content that's been validated
  - Dropping this requirement = potentially rebroadcasting invalid / malicious content into the network
- Validate content in a non-blocking manner
  - Dropping this requirement = not possible, we need to validate content in a non-blocking manner, since content validation might require dropping back into the same overlay network to fetch another piece of data
- Collect all validated content and offer it in a single message
  - Dropping this requirement = increases network traffic / congestion

imo, the last one is the best option to drop. It's possible I'm missing something here. Let me know what you think? Is there an alternative workaround here that can satisfy all three requirements?

Reply:
I get that it's more complex, but I don't understand why it would be impossible, to collect all the validated content into a batch. When all of the content has been found valid or invalid, then it could all be sent all together, right?

FWIW, I think it would be fine to punt this into an issue for performance improvement.

Reply:
I think it is fine to block the thread that is validating the content until all content is validated (we can use different threads in parallel for validating one content item and wait for all threads to finish) and then pass only the valid content in batch to `propagate_gossip` as Jason mentioned.